Docs for specific coalescing. (#5886)
carbon-language/carbon-lang: toolchain/docs/coalesce_generic_lowering.md

# Coalescing generic functions emitted when lowering to LLVM IR

<!--
Part of the Carbon Language project, under the Apache License v2.0 with LLVM
Exceptions. See /LICENSE for license information.
SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
-->

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [Problem details](#problem-details)
    -   [SemIR representation and when to do coalescing](#semir-representation-and-when-to-do-coalescing)
    -   [Recursion and strongly connected components](#recursion-and-strongly-connected-components)
    -   [Complexity](#complexity)
    -   [Canonical specific to use](#canonical-specific-to-use)
-   [Algorithm details](#algorithm-details)
-   [Alternatives considered](#alternatives-considered)
    -   [Coalescing in the front-end vs back-end?](#coalescing-in-the-front-end-vs-back-end)
    -   [When to do coalescing in the front-end?](#when-to-do-coalescing-in-the-front-end)
    -   [Compile-time trade-offs](#compile-time-trade-offs)
-   [Opportunities for further improvement](#opportunities-for-further-improvement)

<!-- tocstop -->

## Overview

> When lowering Carbon generics to LLVM, it is possible we emit duplicate LLVM-IR

Reviewer: ```suggestion
When lowering Carbon generics to LLVM, it is possible we emit duplicate LLVM IR
```

Nit: this doc switches between "LLVM-IR" and "LLVM IR", and I think we usually don't hyphenate, so suggesting that direction.

functions. This document describes the algorithm implemented in

> [lowering](lower.md) for determining when and which generated specifics, while

Reviewer: Perhaps add a mention/link in lower.md to this doc, to help navigation?

FWIW, I'm also thinking we should probably create a `docs/lower/` directory similar to `docs/check/`, but fine splitting that out.

Author: Agreed, but leaving this out of this PR.

different at the Carbon language level, can be coalesced into a single one when
lowering Carbon’s intermediate representation (_SemIR_), to
[LLVM IR](https://llvm.org/docs/LangRef.html).

> The overall goal of this optimization is to avoid generating duplicate LLVM-IR

Reviewer: ```suggestion
The overall goal of this optimization is to avoid generating duplicate LLVM IR
```

code where it is easy to determine this from the front-end. Such an optimization
needs to be done after specialization, but there is some flexibility in when to
do it afterwards: before lowering, through analysis of SemIR or during/after
lowering.

The goal of this doc is to describe the algorithm implemented in

> [https://github.com/carbon-language/carbon-lang/pull/5314](https://github.com/carbon-language/carbon-lang/pull/5314),

Reviewer: ```suggestion
[specifics_coalescer](/toolchain/lower/specific_coalescer.h),
```

Suggesting to link the file rather than the PR because, as the implementation shifts it should make it easier to find.

Note, you could also add a mention in the header to this doc for details (but probably in `https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/coalesce_generic_lowering.md` form).

from putting it into context, to the overall goal, the challenges and where
there is still room for improvement in subsequent iterations.

Determining the impact on compile-time is beyond the scope of this document, but
an important problem to follow up on.

## Problem details

In order to determine if two specific functions are equivalent, and a single one
of them can be used instead of the other, the following need to be considered as

> part of the algorithm and its implementation.

Reviewer: When I saw "problem details", I was expecting only a description of particular facets of the problem. But what it seems you do instead is get into both the problem, and the solution. Maybe "Design details" instead?

> ### SemIR representation and when to do coalescing

Reviewer: ```suggestion
### SemIR representation and why to coalesce during lowering
```

At the very end, you state "The algorithm described below does coalescing of specifics during lowering." I'd suggest leading a little more with this

In SemIR, a specific function is defined by an unique tuple: <`function_id`,

> `specific_id`>. There is a single in-memory representation of a generic

Reviewer: Nit: suggesting parens to wrap a tuple format, just a couple different ways.

```suggestion
In SemIR, a specific function is defined by an unique tuple: `(function_id, specific_id)`. There is a single in-memory representation of a generic
```

```suggestion
In SemIR, a specific function is defined by an unique tuple: `(<function_id>, <specific_id>)`. There is a single in-memory representation of a generic
```

function’s body (not one for each specific), where the instructions that are
different between specifics can be determined, on-demand, based on a given
`specific_id`. Hence, determining if two specifics are equivalent needs to

> analyze if these specific-dependent instructions are equivalent at the LLVM-IR

Reviewer: ```suggestion
analyze if these specific-dependent instructions are equivalent at the LLVM IR
```

level. This can only be determined after the eval phase is complete and using
information on how Carbon types map to `llvm::Type`s.

The algorithm described below does coalescing of specifics during lowering. For

> alternatives considered, see [the section below](#alternatives-considered).

Reviewer: ```suggestion
The algorithm described below does coalescing of specifics during lowering. Also
see [alternatives considered](#alternatives-considered).
```

Suggesting to use the section title as the link, also because it's not the immediate next section.

> ### Recursion and strongly connected components

Reviewer: ```suggestion
### Recursion and strongly connected components (SCCs)
```

You introduce the abbreviation below, but consider starting here since it's the first mention... that'll mean it also shows up in the table of contents for a quick skim

> Comparing if two different specific functions contain (access, invoke, etc) the

Reviewer: ```suggestion
Comparing if two different specific functions contain (access, invoke, etc.) the
```

> same specific-dependent instruction is not straight forward when recursion is

Reviewer: ```suggestion
same specific-dependent instruction is not straightforward when recursion is
```

involved. The simplest example is when A and B each are recursive functions, and

> are equivalent. The check “are A and B equivalent” needs to use that as the

Reviewer: ```suggestion
are equivalent. The check "are A and B equivalent" needs to use that as the
```

Nit: suggesting to avoid curly quotes (the difference is subtle, I'd rather let md decide)

starting assumption, and when a call to A and B respectively are found in each,
then the conclusion is that they are equivalent. In practice this requires

> comparison of `specific_id`s, which in SemIR are distinct.

Reviewer: I'm trying to figure out what this paragraph is trying to say. Part of my confusion may be where you say "needs to use that", I'm not sure what "that" is.

Does something like the following work for you?

```suggestion
are equivalent. The check "are A and B equivalent" needs to start by assuming they are equivalent, and when a self-recursive call is found in each, that call is still equivalent. In practice this requires
comparison of `specific_id`s, which in SemIR are distinct.
```

> In the general case, this analysis needs to analyze the call-graph for all

Reviewer: ```suggestion
In the general case, this analysis needs to analyze the call graph for all
```
Nit: I think call graph isn't normally hyphenated (?)

functions and build strongly connected components (SCCs). Either the call graph
is created before lowering, or it is built while lowering, and in a
post-processing phase we can conclude equivalence and simplify the (read: delete

> unnecessary) emitted LLVM IR. The current implementation does the latter.

Reviewer: For me I always have trouble figuring out whether "a, or b, and c" means "(a or b) and c" vs "a or (b and c)". What do you think of the below phrasing?

```suggestion
functions and build strongly connected components (SCCs). The call graph could either be created before lowering or built while lowering. The current implementation does the latter, and in a
post-processing phase we can conclude equivalence and simplify the emitted LLVM IR by deleting unnecessary parts.
```

A non-viable option is building the call graph based on the information “what

> are all call sites of myself, where I am a specific function”, because this

Reviewer: ```suggestion
A non-viable option is building the call graph based on the information "what
are all call sites of myself, where I am a specific function", because this
```

Curly quotes

information is not available until processing the function bodies of all
specific functions. This is an optimization done so that the definition of a
specific isn’t emitted until a use of it is found. Building that information
would duplicate all the lowering logic, minus the LLVM IR creation.

> ### Complexity

Reviewer: Since you're describing this a bit in form of design, maybe "Function fingerprints"?

```suggestion
### Function fingerprints
```

Even with limiting the comparison of specific functions to those defined from
the same generic, a comparison algorithm would still end up with quadratic
complexity in the number of specifics for that generic.

We define two fingerprints for each specific: the first fingerprint
`common_fingerprint` includes specific-dependent information that does not
include `specific_id` information (this would only be discoverable as equivalent
as part of an equivalence SCC), while the second fingerprint,
`specific_fingerprint`, includes all specific-dependent information. As such,

> two specific functions are not equivalent if their `common_fingerprint` differs.

Reviewer: Perhaps use bullets and highlight `common_fingerprint` for how it differs? e.g.:

```suggestion
We define two fingerprints for each specific:

1. `specific_fingerprint`: Includes all specific-dependent information.
2. `common_fingerprint`: Includes the same except for `specific_id` information, which is only known after building an equivalence SCC.

As such, two specific functions are not equivalent if their `common_fingerprint` differs.
```

Author: I left in a longer explanation after the bullet points, for clarity.

Reviewer: At present, it looks like you have the "As such, two specific functions ..." as part of the second bullet. I'll still suggest putting that in as its own paragraph -- it's contrasting the two fingerprints, and it's also a bit long. :)

Two specific functions are equivalent if their `specific_fingerprint`s are
equal. If the `common_fingerprint`s are equal but the `specific_fingerprint`s
are not, the two functions may still be equivalent.

> Ideally, the `specific_fingerprint` can be used as a unique hash and used to

Reviewer: ```suggestion
Ideally, the `specific_fingerprint` can be used as a unique hash to
```

nit: You already say "used" here

first coalesce all specific functions with this same fingerprint, with no
additional checks. Then, all remaining functions may use the
`common_fingerprint` as another unique hash to group remaining potential

> candidates for coalescing. Then, only those with this same second hash are

Reviewer: ```suggestion
candidates for coalescing. Then, only those with this same `common_fingerprint` are
```

processed in a quadratic pass walking all calls instructions and comparing if
the `specific_id` information is equivalent. This optimization is not currently

> implemented.

Reviewer: I think parts of this are implemented, is the below rephrasing accurate as to what you mean isn't currently implemented?

```suggestion
the `specific_id` information is equivalent. The optimization of clustering by `common_fingerprint` is not currently
implemented.
```

Author: None of the clusterings is currently done.

Reviewer: Simpler suggestion then:

```suggestion
the `specific_id` information is equivalent. These optimizations are not currently
implemented.
```

Note that, if we were to extend the algorithm to do compile-time ICF (Identical
Code Folding), that is define a fingerprint for all functions (not only
specifics of the same generic), based on all their instructions (not only their
specific-dependent ones), the complexity would be much higher, and so would the

> compile-time cost.

Reviewer: I'm trying to figure out what you're trying to say here... Perhaps you could rephrase a little? In what way do you mean "the complexity would be much higher"? Are you specifically referring to lld's ICF implementation?

Perhaps this belongs under "Alternatives considered" rather than here?

Author: I was referring to doing coalescing for any functions in the same FileContext, regardless if they come from the same generic. The function fingerprint would then include all instructions, not only those specific-dependent ones; and to avoid linker issues, you'd likely need to keep the original function definitions and replace the body with a call to the canonical function chosen during coalescing.

This was meant to point out what this algorithm does not do and that it's likely not worth doing without putting too many details in :-).

Reviewer: For me here, an alternative would emphasize that it's something not intended as future work. In my head, this could be an alternative like:

> ### Coalesce duplicate non-specific functions
> 
> We could coalesce duplicate functions in non-specific cases, similar to lld's [Identical Code Folding](https://lld.llvm.org/NewLLD.html#glossary) or LLVM's [MergeFunctions pass](https://llvm.org/docs/MergeFunctions.html). This would require fingerprinting all instructions in all functions, whereas specific coalescing can focus on cases that only Carbon's front-end knows about. Carbon would also be restricted to coalescing functions in a single compilation unit, which would require replacing function definitions that allow external calls with a placeholder that calls the coalesced definition. We don't expect sufficient advantages over existing support.

Would that get at what you're trying to address here? You could still link here, e.g. `Note that this does not [coalesce non-specifics](#coalesce-duplicate-non-specific-functions)`.

Author: SG, thanks!

### Canonical specific to use

For determining the canonical specific to use, we use a
[disjoint set](https://en.wikipedia.org/wiki/Disjoint-set_data_structure).

## Algorithm details

Below is a pseudocode of the existing algorithm in
`toolchain/lower/specific_coalescer.*`.

The implementation has been merged in

> [https://github.com/carbon-language/carbon-lang/pull/5314](https://github.com/carbon-language/carbon-lang/pull/5314)

Reviewer: My above note about linking PR versus code would apply here too. :)

At the top level, the current algorithm first generates all function
definitions, and once this is complete, it performs the logic to coalesce
specifics and delete the redundant LLVM function definitions.

```none
LowerToLLVM () {
  for all non_generic_functions
    CreateLLVMFunctionDefinition (function, no_specific_id);
  PerformCoalescingPostProcessing ();
}
```

The lowering starts with all non-generic functions. While lowering these, when
calls to specifics are encountered, it also generates definitions for those
specific functions.

For each lowered specific function definition, we create the
`SpecificFunctionFingerprint`, which includes the two fingerprints (hashes)

> defined [above](#complexity), and a list of calls to other specific functions.

Reviewer: ```suggestion
`SpecificFunctionFingerprint`, which includes the [two fingerprints](#complexity), and a list of calls to other specific functions.
```

```none
CreateLLVMFunctionDefinition (function, specific_id) {
   Step1: Build llvm::Function*: emit LLVM IR for each SemIR instruction.

   Step2: If the instruction is specific-dependent, hash it and add to its `common_fingerprint`

   Step3: When finding a call to a generic, with a defined type (that is a call to a specific),

>     a) create a definition for this specific_id if it doesn't exist:

Reviewer: Nit -- you might consider consistent capitalization between "Step3: When" and "a) create", either capitalizing both when starting an item in a list or neither

      CreateLLVMFunctionDefinition (function, specific_id);
    b) hash the specific_id to the current function's `specific_fingerprint`

>     c) add the non-hashed specific_id to list of calls performed

Reviewer: It's doing all of these steps for each instruction, right? Maybe:

```suggestion
  For each SemIR instruction in the function:
  
    Step 1: Emit LLVM IR for the instruction

    Step 2: If the instruction is specific-dependent, hash it and add to its `common_fingerprint`

    Step 3: When finding a call to a specific,
      a) create a definition for this specific_id if it doesn't exist:
        CreateLLVMFunctionDefinition (function, specific_id);
      b) hash the specific_id to the current function's `specific_fingerprint`
      c) add the non-hashed specific_id to list of calls performed
```

(note, also suggesting trimming down the "call to a generic, with a defined type (that is a call to a specific)" to just "call to a specific" -- it feels like that's maybe extra detail that's not needed here, similar to how this won't be recursive)

}
```

> The logic that performs the actual coalescing, first checks if the LLVM function

Reviewer: ```suggestion
The logic that performs the actual coalescing analyzes all specifics. For each pair, it first checks if the LLVM function
```

types match (using a third hash-like fingerprint: `function_type_fingerprint`
for storage optimization), then if these are equivalent based on the
`SpecificFunctionFingerprint`. For each pair of equivalent functions found (in a
callgraph SCC), the uses of one of the functions will be replaced with the

> canonical one, while the other definition will ultimately be deleted.

Reviewer: ```suggestion
callgraph SCC), one function will be marked non-canonical: its uses are replaced with the canonical one and its definition will ultimately be deleted.
```

Suggesting a slight rephrasing here, to make it clear "other definition"  means "the replaced one's definition"

```none
PerformCoalescingPostProcessing () {
  for each two specifics of the same generic {
    if function_type_fingerprints differ {
      track as non-equivalent
      continue
    }

    add the two specifics to assumed equivalent specifics list
    if (CheckIfEquivalent(two specifics, assumed equivalent specifics list)) {
      for each two equivalent specifics found {
        find the canonical specific & mark the duplicates for replacement/deletion
    }
  }
  replace all duplicate specifics with the respective canonical specifics
  and delete all replaced LLVM function definitions.
}

```

The equivalence check for specifics based on the constructed
`SpecificFunctionFingerprint` can make an early non-equivalence determination
based on the `common_fingerprint`s, and an early equivalence determination based
on the `specific_fingerprint`s. Otherwise, it uses the call list and recurses to
make the determination for all functions in the SCC call graph (in practice the
implementation uses a worklist to avoid the recursion).

```none

> CheckIfEquivalent(two specifics, mutable list of assumed equivalent specifics) -> bool {

Reviewer: Nit: might want to make naming of "assumed equivalent specifics list" versus "mutable list of assumed equivalent specifics" match

>   if common_fingerprints are non-equal or already tracked as non-equivalent {

Reviewer: What's the equivalent code here? This doesn't look like it maps to the implementation in a way that I'm expecting.

Author: The corresponding code is at specific_coalescer.cpp:204.
I'm taking a look at why I removed the check for "already tracked as non-equivalent".

Reviewer: Can a pair be added to the worklist if they're already tracked as non-equivalent? Inside `AreFunctionBodiesEquivalent` it looks like they won't be added to the worklist unless they're not known. The only pair not examined this way are the inputs, which maybe that's more an issue for `CoalesceEquivalentSpecifics` if it can traverse specifics that were already examined as part of `AreFunctionBodiesEquivalent`?

Author: In the caller of `AreFunctionBodiesEquivalent`, it can only end up looking at a pair that was added to non-equivalent, if those were found by traversing the call graph with a different starting pair. So it's possible, but at this point I don't think it's worth doing a map lookup to check every pair upfront for when that happens. I'm leaving it like this for now. The docs are updated to reflect the current code.

    track as non-equivalent
    return false
  }
  if specific_fingerprints are equal {
    track as equivalent
    return true
  }
  if already tracked as equivalent or assumed equivalent {
    return true
  }

>   for each of the calls in each of the specifics {

Reviewer: Suggesting to match the way you talk about this below.
```suggestion
  for each <pair of calls> in each of the specifics {
```

      continue
    }
    if the functions called are already non-equivalent {
      return false
    }
    add <pair of calls> to assumed equivalent specifics
    if !CheckIfEquivalent(specifics in <pair of calls>, assumed equivalent specifics) {
      return false;
    }
  }
}
```

## Alternatives considered

### Coalescing in the front-end vs back-end?

An alternative considered was not doing any coalescing in the front-end and
relying on LLVM to make the analysis and optimization. The current choice was
made based on the expectation that such an
[LLVM pass](https://llvm.org/docs/MergeFunctions.html) would be more costly in
terms of compile-time. The relative cost has not yet been evaluated.

### When to do coalescing in the front-end?

The analysis and coalescing could be done prior to lowering, after
specialization. The advantage of that choice would be avoiding to lower
duplicate LLVM functions and then removing the duplicates. The disadvantage of
that choice would be duplicating much of the lowering logic, currently necessary
to make the equivalence determination.

### Compile-time trade-offs

Not doing any coalescing is also expected to increase compile time more than

> performing the analysis and deduplication. This can be evaluated in practice and

Reviewer: We're concerned about compile time here, but maybe we can specifically call out the time spent by LLVM versus the front-end time? e.g., does this work for you:

```suggestion
Not doing any coalescing is also expected to increase the back-end codegen time more than
performing the analysis and deduplication. This can be evaluated in practice and
```

the feature disabled if found to be too costly.

## Opportunities for further improvement

The current implemented algorithm can be improved with at least the following:

-   The `specific_fingerprint` can be used to already bucket specifics that can
    be coalesced right away.
-   The remaining ones can be pre-bucketed such that only the specifics with the
    same `common_fingerprint` have their list of calls further compared (linear
    in the number of specific calls inside the functions) to determine SCCs that
    may be equivalent.

This should reduce the complexity from the current N^2, with N=number of

> specifics for a generic, to M^2, with M being the number of specifics for a

Reviewer: ```suggestion
This should reduce the complexity from the current O(N^2), with N=number of
specifics for a generic, to O(M^2), with M being the number of specifics for a
```
